app/src/util_ios.mm

```diff
@@ -76,41 +158,63 @@ - (BOOL)application:(UIApplication *)application
 @end
 
+@implementation UIApplication (FirebaseAppDelegateSwizzling)
+
++ (void)load {
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    Class uiApplicationClass = [UIApplication class];
+    SEL originalSelector = @selector(setDelegate:);
+    Method originalMethod = class_getInstanceMethod(uiApplicationClass, originalSelector);
+
+    if (!originalMethod) {
+      NSLog(@"Firebase Error: Original [UIApplication setDelegate:] method not found for swizzling.");
+      return;
+    }
+
+    // Replace the original method's implementation with Firebase_setDelegate
+    // and store the original IMP.
+    IMP previousImp = method_setImplementation(originalMethod, (IMP)Firebase_setDelegate);
+    if (previousImp) {
+        g_original_setDelegate_imp = previousImp;
+        NSLog(@"Firebase: Successfully swizzled [UIApplication setDelegate:] and stored original IMP.");
+    } else {
+        // This would be unusual - method_setImplementation replacing a NULL IMP,
+        // or method_setImplementation itself failed (though it doesn't typically return NULL on failure,
+        // it might return the new IMP or the old one depending on versions/runtime).
+        // More robustly, g_original_setDelegate_imp should be checked before use.
+        // For now, this logging indicates if previousImp was unexpectedly nil.
+        NSLog(@"Firebase Error: Swizzled [UIApplication setDelegate:], but original IMP was NULL (or method_setImplementation failed to return the previous IMP).");
+    }
+  });
+}
+
+@end
+
 namespace firebase {
 namespace util {
 
-void ForEachAppDelegateClass(void (^block)(Class)) {
-  unsigned int number_of_classes;
-  Class *classes = objc_copyClassList(&number_of_classes);
-  for (unsigned int i = 0; i < number_of_classes; i++) {
-    Class clazz = classes[i];
-    if (class_conformsToProtocol(clazz, @protocol(UIApplicationDelegate))) {
-      const char *class_name = class_getName(clazz);
-      bool blacklisted = false;
-      static const char *kClassNameBlacklist[] = {
-          // Declared in Firebase Analytics:
-          // //googlemac/iPhone/Firebase/Analytics/Sources/ApplicationDelegate/
-          // FIRAAppDelegateProxy.m
-          "FIRAAppDelegate",
-          // Declared here.
-          "FIRSAMAppDelegate"};
-      for (size_t i = 0; i < FIREBASE_ARRAYSIZE(kClassNameBlacklist); ++i) {
-        if (strcmp(class_name, kClassNameBlacklist[i]) == 0) {
-          blacklisted = true;
-          break;
-        }
-      }
-      if (!blacklisted) {
-        if (GetLogLevel() <= kLogLevelDebug) {
-          // Call NSLog directly because we may be in a +load method,
-          // and C++ classes may not be constructed yet.
-          NSLog(@"Firebase: Found UIApplicationDelegate class %s", class_name);
-        }
-        block(clazz);
+void RunOnAppDelegateClasses(void (^block)(Class)) {
+  if (g_seen_delegate_classes_count > 0) {
```

Comment:
Since a lot of these are running at start up time, any concern about multiple threads having problems here?  Like, can RunOnAppDelegateClasses be called by one thread while another is still iterating over the AppDelegates?